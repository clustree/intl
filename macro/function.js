const {
  getMessageDescriptorValue,
  getMessageDescriptorKey,
  storeMessage,
  buildMacroError
} = require("./helpers");

exports.handleFunction = function handleFunction(p, state) {
  let path = p.parentPath;
  if (path.type !== "CallExpression") {
    throw buildMacroError("`translate` not called as a function", path);
  }
  const args = path.get("arguments");
  const [message, options] = [args[0], args[1]];
  const defaultMessage = getMessageDescriptorValue(message);
  let id = defaultMessage;
  let description = null;
  if (options != null) {
    const optionsObj = options.get("properties").reduce((acc, prop) => {
      const key = getMessageDescriptorKey(prop.get("key"));
      if (!["id", "description"].includes(key)) {
        return acc;
      }
      const value = getMessageDescriptorValue(prop.get("value"));
      acc[key] = value;
      return acc;
    }, {});
    if (optionsObj.id) {
      id = optionsObj.id;
    }
    if (optionsObj.description) {
      description = optionsObj.description;
    }
  }

  // Evaluate the Message Descriptor values, then store it.
  const descriptor = {
    defaultMessage,
    id
  };
  if (description != null) {
    descriptor.description = description;
  }
  storeMessage(descriptor, path, state);
};
