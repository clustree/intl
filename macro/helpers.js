const MESSAGES = Symbol("ClustreeIntlMessages");
exports.MESSAGES = MESSAGES;

const { MacroError } = require("babel-plugin-macros");

function buildMacroError(errorMessage, path) {
  const errorMsg = `[@clustree/intl] ${errorMessage}`;
  if (path != null) {
    const error = path.buildCodeFrameError(errorMsg);
    error.name = "MacroError";
    return error;
  }
  return new MacroError(errorMsg);
}

exports.buildMacroError = buildMacroError;

function evaluatePath(path) {
  const evaluated = path.evaluate();
  if (evaluated.confident) {
    return evaluated.value;
  }

  throw buildMacroError(
    "Messages must be statically evaluate-able for extraction.",
    path
  );
}

exports.getMessageDescriptorKey = function getMessageDescriptorKey(path) {
  if (path.isIdentifier() || path.isJSXIdentifier()) {
    return path.node.name;
  }
  throw buildMacroError("Unable to extract descriptorKey", path);
};

exports.getMessageDescriptorValue = function getMessageDescriptorValue(path) {
  if (path.isJSXExpressionContainer()) {
    path = path.get("expression");
  }

  // Always trim the Message Descriptor values.
  const descriptorValue = evaluatePath(path);

  if (typeof descriptorValue === "string") {
    return descriptorValue.trim();
  }
  return descriptorValue;
};

exports.storeMessage = function storeMessage(descriptor, path, state) {
  const { description, defaultMessage, id = defaultMessage } = descriptor;
  const { file } = state;

  if (!defaultMessage) {
    throw buildMacroError(
      "Message Descriptors require a `defaultMessage`.",
      path
    );
  }

  const messages = file.get(MESSAGES);
  if (messages.has(id)) {
    const existing = messages.get(id);
    if (
      description !== existing.description ||
      defaultMessage !== existing.defaultMessage
    ) {
      throw buildMacroError(
        `Duplicate message id: "${id}", ` +
          "but the `description` and/or `defaultMessage` are different.",
        path
      );
    }
  }

  messages.set(id, { id, description, defaultMessage });
};
