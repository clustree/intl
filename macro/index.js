const { handleFunction } = require("./function");
const { handleJSX } = require("./jsx");
const { writeFileSync } = require("./writeFileSync");
const { MESSAGES } = require("./helpers");
const { createMacro } = require("babel-plugin-macros");
const { default: template } = require("@babel/template");
const p = require("path");

module.exports = createMacro(macro, { configName: "@clustree/intl" });

function macro({ references, state, config, babel }) {
  const { file } = state;
  const { types: t } = babel;

  pre(file);

  let imports = {};

  Object.keys(references).map((ref, i) => {
    for (const referencePath of references[ref] || []) {
      imports[ref] = {
        name: `${ref.toUpperCase()}_${i}`,
        value: t.identifier(referencePath.node.name)
      };
      break;
    }
  });

  for (const referencePath of references.translate || []) {
    handleFunction(referencePath, state);
  }

  for (const referencePath of references.Translate || []) {
    handleJSX(referencePath, state, babel);
  }

  post(file, config, imports);
}

function pre(file) {
  if (!file.has(MESSAGES)) {
    file.set(MESSAGES, new Map());
  }
}

function post(file, config, imports) {
  addImports(file, imports);

  const { basename, filename } = file.opts;

  const messages = file.get(MESSAGES);
  const descriptors = [...messages.values()];
  if (config && config.messagesDir && descriptors.length > 0) {
    // Make sure the relative path is "absolute" before
    // joining it with the `messagesDir`.
    const relativePath = p.join(p.sep, p.relative(process.cwd(), filename));

    const messagesFilename = p.join(
      config.messagesDir,
      p.dirname(relativePath),
      `${basename}.json`
    );

    const messagesFile = JSON.stringify(descriptors, null, 2);

    writeFileSync(messagesFilename, messagesFile);
  }
}

function addImports(file, imports) {
  const IMPORT_STRING = Object.entries(imports)
    .map(([key, { name }]) => `${key} as ${name}`)
    .join(", ");
  const IMPORT_VALUES = Object.entries(imports).reduce(
    (acc, [, { name, value }]) => {
      acc[name] = value;
      return acc;
    },
    {}
  );
  file.path.node.body.unshift(
    template(`import { ${IMPORT_STRING} } from '@clustree/intl';`)(
      IMPORT_VALUES
    )
  );
}
