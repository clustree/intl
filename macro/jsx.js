const {
  getMessageDescriptorValue,
  getMessageDescriptorKey,
  storeMessage,
  buildMacroError
} = require("./helpers");

exports.handleJSX = function handleJSX(p, state, babel) {
  // Ignore closing tag
  if (p.parentPath.type === "JSXClosingElement") {
    return;
  }
  let path = p.parentPath.parentPath;
  //  Path is now the JSXElement <Translate />
  if (path.type !== "JSXElement") {
    throw buildMacroError("`Translate` not called as JSX");
  }
  transformJSX(path, state, babel);
};

const mergeProps = (props, nextProps) => ({
  text: props.text + nextProps.text,
  values: Object.assign({}, props.values, nextProps.values),
  components: Object.assign({}, props.components, nextProps.components),
  formats: props.formats,
  elementIndex: nextProps.elementIndex
});

const initialProps = ({ formats } = {}) => ({
  text: "",
  values: {},
  components: {},
  formats: formats || {}
});

function transformJSX(path, state, babel) {
  const { types } = babel;
  processTranslate(path, types);
  extractDefs(path, state);
}

function processElement(node, props, types) {
  const item = node.openingElement.name;
  let name = item.name;
  const index = elementGenerator(name);
  if (index) {
    name = `${name}-${index}`;
  }
  const selfClosing = node.openingElement.selfClosing;

  props.text += !selfClosing ? `<${name}>` : `<${name}/>`;

  for (const child of node.children) {
    props = processChild(child, props, types);
  }

  if (!selfClosing) {
    props.text += `</${name}>`;
  }

  props.components = Object.assign({}, props.components, {
    [name]: types.objectProperty(
      types.isValidIdentifier(name)
        ? types.identifier(name)
        : types.stringLiteral(name),
      node
    )
  });
  return props;
}

const generatorFactory = () => {
  const data = new Map();
  return key => {
    const value = data.get(key) || 0;
    data.set(key, value + 1);
    return value;
  };
};

let elementGenerator;
let argumentGenerator;

function processChild(node, props, types) {
  let nextProps = initialProps({ formats: props.formats });
  if (types.isJSXExpressionContainer(node)) {
    const exp = node.expression;

    if (types.isStringLiteral(exp)) {
      nextProps.text += exp.value;
    } else if (types.isTemplateLiteral(exp)) {
      const parts = [];

      exp.quasis.forEach((item, index) => {
        parts.push(item);

        if (!item.tail) {
          parts.push(exp.expressions[index]);
        }
      });

      parts.forEach(item => {
        if (types.isTemplateElement(item)) {
          nextProps.text += item.value.raw;
        } else {
          const name = item.name;
          const key = item;
          nextProps.text += `{${name}}`;
          nextProps.values[name] = types.objectProperty(key, item);
        }
      });
    } else if (types.isJSXElement(exp)) {
      nextProps = processElement(exp, nextProps);
    } else {
      const name = types.isIdentifier(exp) ? exp.name : argumentGenerator();
      const key = types.isIdentifier(exp) ? exp : types.numericLiteral(name);
      nextProps.text += `{${name}}`;
      // FIXME: This is a workaround that helps Translate work with nested tags
      if (types.isIdentifier(exp)) {
        nextProps.values[name] = types.objectProperty(
          key,
          types.identifier(name)
        );
      } else {
        nextProps.values[name] = types.objectProperty(key, exp);
      }
    }
  } else if (types.isJSXElement(node)) {
    nextProps = processElement(node, nextProps, types);
  } else {
    nextProps.text += node.value;
  }

  return mergeProps(props, nextProps);
}

function processTranslate(path, types) {
  const children = path.node.children;
  if (children.length === 0) {
    return;
  }
  let props = initialProps();
  elementGenerator = generatorFactory();
  argumentGenerator = generatorFactory();
  for (const child of children) {
    props = processChild(child, props, types);
  }

  // replace whitespace before/after newline with single space
  const nlRe = /\s*(?:\r\n|\r|\n)+\s*/g;
  // remove whitespace before/after tag
  const nlTagRe = /(?:(>)(?:\r\n|\r|\n)+\s+|(?:\r\n|\r|\n)+\s+(?=<))/g;

  const text = props.text
    .replace(nlTagRe, "$1")
    .replace(nlRe, " ")
    .trim();

  const values = Object.values(props.values);
  const components = Object.values(props.components);

  const attributes = path.node.openingElement.attributes.slice();
  attributes.push(
    types.JSXAttribute(
      types.JSXIdentifier("defaultMessage"),
      types.StringLiteral(text)
    )
  );
  if (values.length) {
    attributes.push(
      types.JSXAttribute(
        types.JSXIdentifier("values"),
        types.JSXExpressionContainer(types.objectExpression(values))
      )
    );
  }
  if (components.length) {
    attributes.push(
      types.JSXAttribute(
        types.JSXIdentifier("components"),
        types.JSXExpressionContainer(types.objectExpression(components))
      )
    );
  }

  path.replaceWith(
    types.JSXElement(
      types.JSXOpeningElement(
        types.JSXIdentifier("Translate"),
        attributes,
        true
      ),
      null,
      []
    )
  );
}

const DESCRIPTOR_PROPS = new Set(["id", "description", "defaultMessage"]);

function createMessageDescriptor(propPaths) {
  return propPaths.reduce((hash, [keyPath, valuePath]) => {
    const key = getMessageDescriptorKey(keyPath);

    if (DESCRIPTOR_PROPS.has(key)) {
      hash[key] = valuePath;
    }

    return hash;
  }, {});
}

function extractDefs(path, state) {
  const attributes = path
    .get("openingElement")
    .get("attributes")
    .filter(attr => attr.isJSXAttribute());

  if (
    attributes.find(attribute => attribute.node.name.name === "allowDynamic")
  ) {
    // Ignore this Tag as it is dynamic
    return;
  }

  let descriptor = createMessageDescriptor(
    attributes.map(attr => [attr.get("name"), attr.get("value")])
  );

  // In order for a default message to be extracted when
  // declaring a JSX element, it must be done with standard
  // `key=value` attributes. But it's completely valid to write
  // `<Translate id={dynamicId} />`, because it will be
  // skipped here and extracted elsewhere. The descriptor will
  // be extracted only if a `defaultMessage` prop exists.
  if (descriptor.defaultMessage) {
    // Evaluate the Message Descriptor values in a JSX
    // context, then store it.
    descriptor = evaluateMessageDescriptor(descriptor);

    storeMessage(descriptor, path, state);

    // Remove description since it's not used at runtime.
    attributes.some(attr => {
      const ketPath = attr.get("name");
      if (getMessageDescriptorKey(ketPath) === "description") {
        attr.remove();
        return true;
      }
      return false;
    });
  } else {
    const error = path.buildCodeFrameError(
      "[Clustree Intl] No message defined in Translate."
    );
    error.name = "MacroError";
    throw error;
  }
}

function evaluateMessageDescriptor({ ...descriptor }) {
  Object.keys(descriptor).forEach(key => {
    const valuePath = descriptor[key];

    descriptor[key] = getMessageDescriptorValue(valuePath);
  });

  return descriptor;
}
