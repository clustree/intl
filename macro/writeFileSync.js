const path = require("path");
const fs = require("fs");
const { sync: mkdirpSync } = require("mkdirp");

exports.writeFileSync = function writeFileSync(messagesFilename, messagesFile) {
  mkdirpSync(path.dirname(messagesFilename));
  fs.writeFileSync(messagesFilename, messagesFile);
};
