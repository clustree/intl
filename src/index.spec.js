/* eslint-env jest */
import React from "react";
import renderer from "react-test-renderer";
import { Provider, translate, Translate } from "../macro";

const smallStrong = {
  small: <small />,
  strong: <strong />
};

function checkSnapshot(elt, { locale = "en", messages } = {}) {
  expect(
    renderer
      .create(
        <Provider locale={locale} messages={messages} defaultLocale="en">
          {elt}
        </Provider>
      )
      .toJSON()
  ).toMatchSnapshot();
}

describe("Translate", () => {
  it("Translates with JSX in values", () => {
    checkSnapshot(
      <Translate
        defaultMessage={`{nb, plural,
          one {<strong>singular</strong><small>jsx</small>}
          other {<strong>plural</strong><small>jsx</small>}
        }`}
        values={{ nb: 5 }}
        components={smallStrong}
      />
    );

    checkSnapshot(
      <Translate>
        <strong>bold</strong>
        <small>tiny</small>
      </Translate>
    );

    checkSnapshot(
      <Translate>
        <strong>bold</strong>
        <small>tiny</small>
      </Translate>,
      {
        locale: "fr",
        messages: {
          "<strong>bold</strong><small>tiny</small>":
            "<small>Tres petit</small><strong>gras</strong>"
        }
      }
    );
  });
});

describe("Translate", () => {
  it("Escapes JSX in values", () => {
    const strong = "<strong/>";
    checkSnapshot(<Translate components={smallStrong}>bob {strong}</Translate>);
    const details = "<details/>";
    checkSnapshot(
      <Translate components={smallStrong}>bob {details}</Translate>
    );

    checkSnapshot(<Translate components={smallStrong}>bob</Translate>);

    checkSnapshot(
      <Translate
        values={{ jsx: <strong /> }}
        defaultMessage="{jsx} in string"
      />
    );
  });
});

describe("translate", () => {
  it("gives the correct result", () => {
    expect(
      translate("test {value}", {
        values: { value: 4 }
      })
    ).toEqual("test 4");
  });
});
