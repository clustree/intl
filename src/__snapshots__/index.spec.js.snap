// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynamic translations Translates with dynamic key 1`] = `
Array [
  "custom1 name",
  "custom2 name",
]
`;

exports[`Translate Escapes JSX in values 1`] = `
Array [
  "bob ",
  "<strong/>",
  "",
]
`;

exports[`Translate Escapes JSX in values 2`] = `
Array [
  "bob ",
  "<details/>",
  "",
]
`;

exports[`Translate Escapes JSX in values 3`] = `"bob"`;

exports[`Translate Escapes JSX in values 4`] = `
Array [
  "",
  <strong />,
  " in string",
]
`;

exports[`Translate Renders null on Error 1`] = `null`;

exports[`Translate Renders null on Error 2`] = `null`;

exports[`Translate Renders null on Error 3`] = `null`;

exports[`Translate Translates with JSX in values 1`] = `
Array [
  "",
  <strong>
    plural
  </strong>,
  "",
  <small>
    jsx
  </small>,
  "",
]
`;

exports[`Translate Translates with JSX in values 2`] = `
Array [
  "",
  <strong>
    bold
  </strong>,
  "",
  <small>
    tiny
  </small>,
  "",
]
`;

exports[`Translate Translates with JSX in values 3`] = `
Array [
  "",
  <small>
    Tres petit
  </small>,
  "",
  <strong>
    gras
  </strong>,
  "",
]
`;
